File: services/relaymock/relaymock/routers/deliveries.py
from __future__ import annotations

import sqlite3
from datetime import timedelta

from fastapi import APIRouter, Depends, Response, status
from fastapi.security import HTTPAuthorizationCredentials, HTTPBearer

from ..api_contract import error_responses
from ..auth import AuthContext, get_auth_context, get_database
from ..database import Database
from ..schemas import FileDeliveryEventIn, FileDeliveryOut
from ..services import http_error
from ..utils import from_iso, hash_token, new_id, to_iso, utc_now

router = APIRouter(tags=["file deliveries"])
delivery_bearer = HTTPBearer(auto_error=False, scheme_name="DeliveryBearer")


def _out(row: sqlite3.Row) -> FileDeliveryOut:
    return FileDeliveryOut(
        id=row["id"],
        push_id=row["push_id"],
        file_id=row["file_id"],
        destination_device_id=row["destination_device_id"],
        state=row["state"],
        created_at=row["created_at"],
        updated_at=row["updated_at"],
        notified_at=row["notified_at"],
        fetching_at=row["fetching_at"],
        cached_at=row["cached_at"],
        failed_at=row["failed_at"],
        missed_at=row["missed_at"],
        failure_code=row["failure_code"],
        attempt_count=row["attempt_count"],
    )


def ensure_file_deliveries(conn: sqlite3.Connection, push: sqlite3.Row, now: str) -> None:
    if push["type"] != "file" or not push["file_id"]:
        return
    if push["target_kind"] == "device":
        devices = conn.execute(
            "SELECT id FROM devices WHERE id = ? AND user_id = ? AND revoked_at IS NULL",
            (push["target_device_id"], push["user_id"]),
        ).fetchall()
    elif push["target_kind"] == "all_devices":
        devices = conn.execute(
            "SELECT id FROM devices WHERE user_id = ? AND revoked_at IS NULL ORDER BY id",
            (push["user_id"],),
        ).fetchall()
    else:
        devices = conn.execute(
            "SELECT id FROM devices WHERE user_id = ? AND id != ? AND revoked_at IS NULL ORDER BY id",
            (push["user_id"], push["source_device_id"]),
        ).fetchall()
    for device in devices:
        conn.execute(
            """
            INSERT OR IGNORE INTO file_deliveries(
                id, user_id, push_id, file_id, destination_device_id, state,
                created_at, updated_at, attempt_count
            ) VALUES (?, ?, ?, ?, ?, 'pending', ?, ?, 0)
            """,
            (new_id("fdl"), push["user_id"], push["id"], push["file_id"], device["id"], now, now),
        )


def issue_delivery_token(conn: sqlite3.Connection, delivery_id: str, token: str) -> bool:
    now = utc_now()
    result = conn.execute(
        """
        UPDATE file_deliveries
        SET ack_token_hash = ?, ack_token_expires_at = ?,
            state = CASE WHEN state = 'pending' THEN 'notified' ELSE state END,
            notified_at = COALESCE(notified_at, ?), updated_at = ?,
            attempt_count = attempt_count + 1
        WHERE id = ? AND state IN ('pending', 'notified', 'failed_retryable')
        """,
        (hash_token(token), to_iso(now + timedelta(days=1)), to_iso(now), to_iso(now), delivery_id),
    )
    return result.rowcount == 1


def mark_undelivered_missed(conn: sqlite3.Connection, file_id: str, reason: str, now: str) -> None:
    conn.execute(
        """
        UPDATE file_deliveries
        SET state = 'missed', updated_at = ?, missed_at = ?, failure_code = ?
        WHERE file_id = ? AND state != 'cached'
        """,
        (now, now, reason, file_id),
    )


@router.get(
    "/v1/files/{file_id}/deliveries",
    response_model=list[FileDeliveryOut],
    responses=error_responses(401, 404),
)
def list_file_deliveries(
    file_id: str,
    auth: AuthContext = Depends(get_auth_context),
    database: Database = Depends(get_database),
) -> list[FileDeliveryOut]:
    with database.connection() as conn:
        owned = conn.execute(
            "SELECT id FROM files WHERE id = ? AND user_id = ?", (file_id, auth.user_id)
        ).fetchone()
        if owned is None:
            raise http_error(status.HTTP_404_NOT_FOUND, "file_not_found", "File not found.")
        rows = conn.execute(
            "SELECT * FROM file_deliveries WHERE file_id = ? AND user_id = ? ORDER BY destination_device_id, id",
            (file_id, auth.user_id),
        ).fetchall()
    return [_out(row) for row in rows]


@router.post(
    "/v1/file-deliveries/{delivery_id}/events",
    response_model=FileDeliveryOut,
    responses=error_responses(401, 403, 409, 410),
)
def acknowledge_file_delivery(
    delivery_id: str,
    body: FileDeliveryEventIn,
    response: Response,
    credentials: HTTPAuthorizationCredentials | None = Depends(delivery_bearer),
    database: Database = Depends(get_database),
) -> FileDeliveryOut:
    if credentials is None or credentials.scheme.lower() != "bearer" or not credentials.credentials:
        raise http_error(status.HTTP_401_UNAUTHORIZED, "delivery_token_required", "A delivery acknowledgement token is required.")
    with database.connection() as conn:
        row = conn.execute(
            "SELECT * FROM file_deliveries WHERE id = ? AND ack_token_hash = ?",
            (delivery_id, hash_token(credentials.credentials)),
        ).fetchone()
        if row is None:
            raise http_error(status.HTTP_403_FORBIDDEN, "invalid_delivery_token", "The delivery acknowledgement token is invalid.")
        now = utc_now()
        if row["ack_token_expires_at"] is None or from_iso(row["ack_token_expires_at"]) <= now:
            raise http_error(status.HTTP_410_GONE, "delivery_token_expired", "The delivery acknowledgement token has expired.")
        if row["state"] == "cached":
            response.headers["Idempotent-Replayed"] = "true"
            return _out(row)
        if row["state"] == "missed":
            raise http_error(status.HTTP_409_CONFLICT, "delivery_missed", "A missed delivery cannot be acknowledged.")
        timestamp = to_iso(now)
        failure_code = body.failure_code if body.state == "failed_retryable" else None
        conn.execute(
            """
            UPDATE file_deliveries SET state = ?, updated_at = ?,
                fetching_at = CASE WHEN ? = 'fetching' THEN COALESCE(fetching_at, ?) ELSE fetching_at END,
                cached_at = CASE WHEN ? = 'cached' THEN COALESCE(cached_at, ?) ELSE cached_at END,
                failed_at = CASE WHEN ? = 'failed_retryable' THEN ? ELSE failed_at END,
                failure_code = CASE WHEN ? = 'failed_retryable' THEN ? WHEN ? = 'cached' THEN NULL ELSE failure_code END
            WHERE id = ?
            """,
            (body.state, timestamp, body.state, timestamp, body.state, timestamp,
             body.state, timestamp, body.state, failure_code, body.state, delivery_id),
        )
        conn.commit()
        updated = conn.execute("SELECT * FROM file_deliveries WHERE id = ?", (delivery_id,)).fetchone()
    return _out(updated)

File: infra/cloudflare/worker/src/deliveries.ts
import { sha256Hex } from "./crypto";
import { bodyJson, json, problem } from "./response";
import { iso } from "./runtime";
import type { AuthContext, Env, FileDeliveryRow, PushRow, Runtime } from "./types";

const ACK_TOKEN_TTL_MS = 24 * 60 * 60 * 1000;

function deliveryOut(row: FileDeliveryRow): Record<string, unknown> {
  return {
    id: row.id,
    push_id: row.push_id,
    file_id: row.file_id,
    destination_device_id: row.destination_device_id,
    state: row.state,
    created_at: iso(row.created_at),
    updated_at: iso(row.updated_at),
    notified_at: iso(row.notified_at),
    fetching_at: iso(row.fetching_at),
    cached_at: iso(row.cached_at),
    failed_at: iso(row.failed_at),
    missed_at: iso(row.missed_at),
    failure_code: row.failure_code,
    attempt_count: Number(row.attempt_count),
  };
}

export async function ensureFileDeliveries(env: Env, push: PushRow, runtime: Runtime): Promise<void> {
  if (push.type !== "file" || !push.file_id) return;
  const targetKind = push.target_kind ?? (push.target_device_id ? "device" : "all_other_devices");
  let devices;
  if (targetKind === "device") {
    devices = await env.DB.prepare("SELECT id FROM devices WHERE id = ? AND user_id = ? AND revoked_at IS NULL")
      .bind(push.target_device_id, push.user_id).all<{ id: string }>();
  } else if (targetKind === "all_devices") {
    devices = await env.DB.prepare("SELECT id FROM devices WHERE user_id = ? AND revoked_at IS NULL ORDER BY id")
      .bind(push.user_id).all<{ id: string }>();
  } else {
    devices = await env.DB.prepare("SELECT id FROM devices WHERE user_id = ? AND id != ? AND revoked_at IS NULL ORDER BY id")
      .bind(push.user_id, push.source_device_id).all<{ id: string }>();
  }
  const now = runtime.now();
  if (devices.results.length === 0) return;
  await env.DB.batch(devices.results.map((device) => env.DB.prepare(`INSERT OR IGNORE INTO file_deliveries
    (id, user_id, push_id, file_id, destination_device_id, state, created_at, updated_at, attempt_count)
    VALUES (?, ?, ?, ?, ?, 'pending', ?, ?, 0)`)
    .bind(runtime.id("fdl"), push.user_id, push.id, push.file_id, device.id, now, now)));
}

export async function issueDeliveryToken(env: Env, deliveryId: string, runtime: Runtime): Promise<{ token: string; expiresAt: number } | null> {
  const token = runtime.token();
  const expiresAt = runtime.now() + ACK_TOKEN_TTL_MS;
  const result = await env.DB.prepare(`UPDATE file_deliveries SET ack_token_hash = ?, ack_token_expires_at = ?,
    updated_at = ?, attempt_count = attempt_count + 1
    WHERE id = ? AND state IN ('pending', 'notified', 'failed_retryable')`)
    .bind(await sha256Hex(token), expiresAt, runtime.now(), deliveryId).run();
  return result.meta.changes === 1 ? { token, expiresAt } : null;
}

export async function markDeliveryNotified(env: Env, deliveryId: string, runtime: Runtime): Promise<void> {
  const now = runtime.now();
  await env.DB.prepare(`UPDATE file_deliveries SET state = 'notified', notified_at = COALESCE(notified_at, ?),
    updated_at = ?, failure_code = NULL WHERE id = ? AND state IN ('pending', 'notified', 'failed_retryable')`)
    .bind(now, now, deliveryId).run();
}

export async function markDeliveryFailed(env: Env, deliveryId: string, failureCode: string, runtime: Runtime): Promise<void> {
  const now = runtime.now();
  await env.DB.prepare(`UPDATE file_deliveries SET state = 'failed_retryable', failed_at = ?, updated_at = ?,
    failure_code = ? WHERE id = ? AND state IN ('pending', 'notified', 'failed_retryable')`)
    .bind(now, now, failureCode.slice(0, 100), deliveryId).run();
}

export async function listFileDeliveries(env: Env, auth: AuthContext, requestId: string, fileId: string): Promise<Response> {
  const owned = await env.DB.prepare("SELECT id FROM files WHERE id = ? AND user_id = ?").bind(fileId, auth.user_id).first();
  if (!owned) return problem(404, "file_not_found", "File not found.", requestId);
  const rows = await env.DB.prepare(`SELECT * FROM file_deliveries WHERE file_id = ? AND user_id = ?
    ORDER BY destination_device_id, id`).bind(fileId, auth.user_id).all<FileDeliveryRow>();
  return json(rows.results.map(deliveryOut), { headers: { "x-request-id": requestId } });
}

export async function handlePublicDeliveryRoute(
  request: Request,
  env: Env,
  requestId: string,
  path: string,
  runtime: Runtime,
): Promise<Response | null> {
  const match = path.match(/^\/v1\/file-deliveries\/([^/]+)\/events$/);
  if (!match || request.method !== "POST") return null;
  const tokenHeader = request.headers.get("authorization") ?? "";
  const token = tokenHeader.match(/^Bearer\s+(.+)$/i)?.[1];
  if (!token) return problem(401, "delivery_token_required", "A delivery acknowledgement token is required.", requestId);
  const row = await env.DB.prepare("SELECT * FROM file_deliveries WHERE id = ? AND ack_token_hash = ?")
    .bind(decodeURIComponent(match[1]), await sha256Hex(token)).first<FileDeliveryRow>();
  if (!row) return problem(403, "invalid_delivery_token", "The delivery acknowledgement token is invalid.", requestId);
  if (row.ack_token_expires_at == null || row.ack_token_expires_at <= runtime.now()) {
    return problem(410, "delivery_token_expired", "The delivery acknowledgement token has expired.", requestId);
  }
  const body = await bodyJson(request, requestId);
  const next = typeof body.state === "string" ? body.state : "";
  if (!['fetching', 'cached', 'failed_retryable'].includes(next)) {
    return problem(422, "invalid_delivery_state", "state must be fetching, cached, or failed_retryable.", requestId);
  }
  if (row.state === "cached") return json(deliveryOut(row), { headers: { "idempotent-replayed": "true", "x-request-id": requestId } });
  if (row.state === "missed") return problem(409, "delivery_missed", "A missed delivery cannot be acknowledged.", requestId);
  const now = runtime.now();
  const failureCode = next === "failed_retryable" && typeof body.failure_code === "string"
    ? body.failure_code.slice(0, 100)
    : null;
  await env.DB.prepare(`UPDATE file_deliveries SET state = ?, updated_at = ?,
    fetching_at = CASE WHEN ? = 'fetching' THEN COALESCE(fetching_at, ?) ELSE fetching_at END,
    cached_at = CASE WHEN ? = 'cached' THEN COALESCE(cached_at, ?) ELSE cached_at END,
    failed_at = CASE WHEN ? = 'failed_retryable' THEN ? ELSE failed_at END,
    failure_code = CASE WHEN ? = 'failed_retryable' THEN ? WHEN ? = 'cached' THEN NULL ELSE failure_code END
    WHERE id = ?`)
    .bind(next, now, next, now, next, now, next, now, next, failureCode, next, row.id).run();
  const updated = await env.DB.prepare("SELECT * FROM file_deliveries WHERE id = ?").bind(row.id).first<FileDeliveryRow>();
  if (!updated) throw new Error("updated delivery is missing");
  return json(deliveryOut(updated), { headers: { "x-request-id": requestId } });
}

export async function markUndeliveredMissed(env: Env, fileId: string, reason: string, runtime: Runtime): Promise<void> {
  const now = runtime.now();
  await env.DB.prepare(`UPDATE file_deliveries SET state = 'missed', updated_at = ?, missed_at = ?, failure_code = ?
    WHERE file_id = ? AND state != 'cached'`).bind(now, now, reason, fileId).run();
}
